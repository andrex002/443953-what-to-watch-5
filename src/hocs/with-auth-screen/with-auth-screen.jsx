import React, {PureComponent, createRef} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {login} from "../../store/api-action";

const withAuthScreen = (Component) => {
  class WithAuthScreen extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        isValidEmail: true,
        isValidPassword: true,
      };

      this._emailRef = createRef();
      this._passwordRef = createRef();

      this._validateEmail = this._validateEmail.bind(this);
      this._validatePassword = this._validatePassword.bind(this);
      this._handleFormSubmit = this._handleFormSubmit.bind(this);
    }

    _validateEmail(email) {
      if (email.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i)) {
        this.setState({
          isValidEmail: true
        });
        return true;
      } else {
        this.setState({
          isValidEmail: false,
        });
        return false;
      }
    }

    _validatePassword(password) {
      if (password) {
        this.setState({
          isValidPassword: true
        });
        return true;
      } else {
        this.setState({
          isValidPassword: false,
        });
        return false;
      }
    }

    _handleFormSubmit(evt) {
      evt.preventDefault();

      const {onSubmit} = this.props;
      const email = this._emailRef.current.value;
      const password = this._passwordRef.current.value;

      if (this._validateEmail(email) && this._validatePassword(password)) {
        onSubmit({
          email: this._emailRef.current.value,
          password: this._passwordRef.current.value
        });
      }
    }

    render() {
      const {isValidEmail, isValidPassword} = this.state;
      const {authorizationStatus} = this.props;
      return (
        <Component {...this.props}
          authorizationStatus={authorizationStatus}
          isValidEmail={isValidEmail}
          isValidPassword={isValidPassword}
          onFormSubmit={this._handleFormSubmit}
          renderEmailInput = {() => {
            return (
              <input ref={this._emailRef} className="sign-in__input" type="email" placeholder="Email address" name="user-email" id="user-email" />
            );
          }}
          renderPasswordInput={() => {
            return (
              <input ref={this._passwordRef} className="sign-in__input" type="password" placeholder="Password" name="user-password" id="user-password" />
            );
          }}
        />
      );
    }
  }

  WithAuthScreen.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    authorizationStatus: PropTypes.string.isRequired
  };

  const mapStateToProps = ({USER}) => ({
    authorizationStatus: USER.authorizationStatus,
  });

  const mapDispatchToProps = (dispatch) => ({
    onSubmit(authData) {
      dispatch(login(authData));
    }
  });

  return connect(mapStateToProps, mapDispatchToProps)(WithAuthScreen);
};

export {withAuthScreen};
export default withAuthScreen;
