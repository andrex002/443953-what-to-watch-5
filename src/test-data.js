export const film = {
  "id": 1,
  "title": "The Grand Budapest Hotel",
  "image": "img/the-grand-budapest-hotel-poster.jpg",
  "previewImage": "img/the-grand-budapest-hotel.jpg",
  "bgImage": "img/the-grand-budapest-hotel-bg.jpg",
  "bgColor": "#ffffff",
  "srcVideo": "https://some-link",
  "video": "https://some-link",
  "previewVideo": "https://some-link",
  "description": "In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.",
  "rating": {
    "score": 8.9,
    "count": 240
  },
  "director": "Wes Andreson",
  "actors": ["Bill Murray", "Edward Norton", "Jude Law", "Willem Dafoe", "Saoirse Ronan"],
  "duration": 99,
  "genre": "Comedy",
  "year": 2014,
  "isFavorite": false
};

export const films = [
  {
    "id": 1,
    "title": "The Grand Budapest Hotel",
    "image": "img/the-grand-budapest-hotel-poster.jpg",
    "previewImage": "img/the-grand-budapest-hotel.jpg",
    "bgImage": "img/the-grand-budapest-hotel-bg.jpg",
    "bgColor": "#ffffff",
    "video": "https://some-link",
    "previewVideo": "https://some-link",
    "description": "In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.",
    "rating": {
      "score": 8.9,
      "count": 240
    },
    "director": "Wes Andreson",
    "actors": ["Bill Murray", "Edward Norton", "Jude Law", "Willem Dafoe", "Saoirse Ronan"],
    "duration": 99,
    "genre": "Comedy",
    "year": 2014,
    "isFavorite": false
  },
  {
    "id": 2,
    "title": "The Grand Budapest Hotel",
    "image": "img/the-grand-budapest-hotel-poster.jpg",
    "previewImage": "img/the-grand-budapest-hotel.jpg",
    "bgImage": "img/the-grand-budapest-hotel-bg.jpg",
    "bgColor": "#ffffff",
    "video": "https://some-link",
    "previewVideo": "https://some-link",
    "description": "In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.",
    "rating": {
      "score": 8.9,
      "count": 240
    },
    "director": "Wes Andreson",
    "actors": ["Bill Murray", "Edward Norton", "Jude Law", "Willem Dafoe", "Saoirse Ronan"],
    "duration": 99,
    "genre": "Comedy",
    "year": 2014,
    "isFavorite": false
  }
];

export const comment = {
  "id": 1,
  "userId": 4,
  "userName": "Kate Muir",
  "rating": 8.9,
  "textComment": "Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.",
  "date": "2019-05-08T14:13:56.569Z"
};

export const comments = [
  {
    "id": 1,
    "userId": 4,
    "userName": "Kate Muir",
    "rating": 8.9,
    "textComment": "Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.",
    "date": "2019-05-08T14:13:56.569Z"
  },
  {
    "id": 2,
    "userId": 4,
    "userName": "Kate Muir",
    "rating": 8.9,
    "textComment": "Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.",
    "date": "2019-05-08T14:13:56.569Z"
  },
];

export const defaultState = {
  DATA: {
    allFilms: films,
    filteredFilms: films,
    promoFilm: film,
    currentFilm: film,
    favoriteFilms: films,
    comments: comments,
    numberFilmsShown: 8,
    isCommentSending: false,
    isCommentSendError: false
  },
  FILTER: {
    activeGenre: `All genres`,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    name: `userName`,
    avatar: `avatar.jpg`
  }
};

export const noop = () => {};