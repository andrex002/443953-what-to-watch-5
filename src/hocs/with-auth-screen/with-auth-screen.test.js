import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureMockStore from 'redux-mock-store';
import {withAuthScreen} from "./with-auth-screen";
import PropTypes from "prop-types";
import {defaultState} from "../../test-data";

const mockStore = configureMockStore()(defaultState);

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

const MockComponentWrapped = withAuthScreen(MockComponent);

it(`withAuthScreen is rendered correctly`, () => {
  const tree = renderer.create((
    <Provider store={mockStore}>
      <MockComponentWrapped>
        <React.Fragment />
      </MockComponentWrapped>
    </Provider>
  )).toJSON();

  expect(tree).toMatchSnapshot();
});
