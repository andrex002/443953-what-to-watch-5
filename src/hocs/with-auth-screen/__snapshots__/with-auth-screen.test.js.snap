// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withAuthScreen is rendered correctly 1`] = `<div />`;
